### DISCUSSION TITLE
JSON Structured logs

### DISCUSSION THREAD

**Discussion by ARobertsCollibra** [qos-ch/logback#667] [Q&A]
Hello Everyone,

I am trying to output logs in a JSON structured format that includes the Key Value Pairs (%kvp).
When I use the normal pattern layout I can get close, but there isn't a good way to convert the KVP to native JSON because the Keys aren't in quotes `"`.  The best I have found is using `%replace` which can get `v1="k1" v2="k2"` to `v1:"k1",v2:"k2"`
```
<pattern>{"timestamp":"%d{yyyyMMdd HH:mm:ss.SSS}","parameters":[%replace(%replace(%kvp{double}){'=',':'}){'" ','",'}],"message":"%msg"}%n</pattern>
```

I started looking at the code base to see how hard it would be to support `%kvp{json}` as an option and came across `ch.qos.logback.classic.encoder.JsonEncoder` which seems like it would do what I need, but there appears to be zero documentation (at least google doesn't find any) and I can't figure out how to define what fields the encoder outputs. 

Any suggestions on the best way to move forward?

Thank you,
Adam

**Comment by ceki** [comment]
@ARobertsCollibra  Hi Adam, 

JsonAppendeer is currently being implemented. See commit 9e833ec858953a  

This issue is being tracked under [LOGBACK-1734](https://jira.qos.ch/browse/LOGBACK-1734)

Documentation should follow shortly

**Reply by ARobertsCollibra** [reply]
Thank you for the quick answer @ceki !

Is there an eta for when [LOGBACK-1734](https://jira.qos.ch/browse/LOGBACK-1734) will be released?   I ask because if it will take a significant length of time I will continue to look into enhancing the PatternEncoder to support a JSON form for `%kvp`.

**Comment by rajivharris** [comment]
Hi @ceki 
I want to log only selected fields in the new JsonEncoder. Not able to find documentation or examples.
How can we select fields and customize filed names?
Any help is appreciated.

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE configuration>

<configuration>
    <import class="ch.qos.logback.classic.encoder.JsonEncoder" />
    <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">    
        <encoder class="ch.qos.logback.classic.encoder.JsonEncoder">
            <!--something like this -->
            <fieldNames>
                <timestamp>timestamp</timestamp>
                <message>message</message>
            </fieldNames>
        </encoder>
               
    </appender>
    <root level="INFO" >        
        <appender-ref ref="STDOUT" />
    </root>

    <logger name="org.springframework" level="INFO" />
    <logger name="org.springframework.web" level="INFO"/>
</configuration>
```

**Reply by ceki** [reply]
Some documentation can be found at https://logback.qos.ch/manual/encoders.html#JsonEncoder
